### DISCUSSION TITLE
Why are some `OpTag`s not available as `NodeHandle`?

### DISCUSSION THREAD

**Discussion by lmondada** [Quantinuum/hugr#427] [Q&A] [answered]
We had introduced a `OpTag::DataflowParent` to refer to nodes with sibling graphs that have an input and output node (useful within DFG rewriting).

I wanted to express this requirement by specifying a `NodeHandle` in my `HugrView` trait, but noticed that it is not available. `HugrView<DfgID>` is too specific. It seems `NodeHandle`s are only available for concrete `OpType`s. Is there a reason for this?

I've gone with `HugrView<Root> where Root: ContainerHandle<ChildrenHandle = DataflowOpID>`, which is esentially equivalent to what I need. Is the `ContainerHandle` trait making `OpTag::DataflowParent` redundant? Can we get rid of redundancy AND keep such requirements reasonable concise?

EDIT: I think this question applies to all "abstract" `OpTag`s, i.e. `OpTag`s that are supersets of `OpType`s.

**Comment by aborgna-q** [comment, marked as answer]
Ideally we'd use `OpTag`s as root node restrictions, but notice that tags are enum values while `NodeHandle`s are types implementing a trait.

We cannot parametrize types with values, so we're restricted to using handle types with associated tags instead.
(Rust supports const generics like `Array<const N: usize>`, but it doesn't allow arbitrary enums like `OpTag`).

We could define more generic handles for each available tag, and `NodeHandle::try_cast` where necessary. I just worry that the explicit calls may get annoying.
